### PR TITLE
python3Packages.django-debug-toolbar: init at 3.2.2

### PR DESCRIPTION
**Pull request by yuuyins**
###### Motivation for this change
I was trying to package https://github.com/vitorfs/parsifal, but it requires `django-debug-toolbar` ( https://github.com/vitorfs/parsifal/blob/dev/requirements/local.txt) which isn't on nixpkgs yet.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

###### Other
On the `checkPhase` I’m getting `django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.`

    ============================= test session starts ==============================
    platform linux -- Python 3.9.6, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
    rootdir: /build/source
    collected 51 items / 6 errors / 45 selected
    
    ==================================== ERRORS ====================================
    __________________ ERROR collecting tests/test_integration.py __________________
    tests/test_integration.py:21: in <module>
        from .views import regular_view
    tests/views.py:1: in <module>
        from django.contrib.auth.models import User
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/models.py:2: in <module>
        from django.contrib.auth.base_user import AbstractBaseUser, BaseUserManager
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/base_user.py:47: in <module>
        class AbstractBaseUser(models.Model):
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/db/models/base.py:103: in __new__
        app_config = apps.get_containing_app_config(module)
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:252: in get_containing_app_config
        self.check_apps_ready()
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:135: in check_apps_ready
        raise AppRegistryNotReady("Apps aren't loaded yet.")
    E   django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
    ____________ ERROR collecting tests/commands/test_debugsqlshell.py _____________
    tests/commands/test_debugsqlshell.py:5: in <module>
        from django.contrib.auth.models import User
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/models.py:2: in <module>
        from django.contrib.auth.base_user import AbstractBaseUser, BaseUserManager
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/base_user.py:47: in <module>
        class AbstractBaseUser(models.Model):
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/db/models/base.py:103: in __new__
        app_config = apps.get_containing_app_config(module)
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:252: in get_containing_app_config
        self.check_apps_ready()
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:135: in check_apps_ready
        raise AppRegistryNotReady("Apps aren't loaded yet.")
    E   django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
    ________________ ERROR collecting tests/panels/test_logging.py _________________
    tests/panels/test_logging.py:9: in <module>
        from ..views import regular_view
    tests/views.py:1: in <module>
        from django.contrib.auth.models import User
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/models.py:2: in <module>
        from django.contrib.auth.base_user import AbstractBaseUser, BaseUserManager
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/base_user.py:47: in <module>
        class AbstractBaseUser(models.Model):
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/db/models/base.py:103: in __new__
        app_config = apps.get_containing_app_config(module)
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:252: in get_containing_app_config
        self.check_apps_ready()
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:135: in check_apps_ready
        raise AppRegistryNotReady("Apps aren't loaded yet.")
    E   django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
    _______________ ERROR collecting tests/panels/test_profiling.py ________________
    tests/panels/test_profiling.py:1: in <module>
        from django.contrib.auth.models import User
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/models.py:2: in <module>
        from django.contrib.auth.base_user import AbstractBaseUser, BaseUserManager
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/base_user.py:47: in <module>
        class AbstractBaseUser(models.Model):
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/db/models/base.py:103: in __new__
        app_config = apps.get_containing_app_config(module)
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:252: in get_containing_app_config
        self.check_apps_ready()
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:135: in check_apps_ready
        raise AppRegistryNotReady("Apps aren't loaded yet.")
    E   django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
    __________________ ERROR collecting tests/panels/test_sql.py ___________________
    tests/panels/test_sql.py:7: in <module>
        from django.contrib.auth.models import User
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/models.py:2: in <module>
        from django.contrib.auth.base_user import AbstractBaseUser, BaseUserManager
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/base_user.py:47: in <module>
        class AbstractBaseUser(models.Model):
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/db/models/base.py:103: in __new__
        app_config = apps.get_containing_app_config(module)
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:252: in get_containing_app_config
        self.check_apps_ready()
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:135: in check_apps_ready
        raise AppRegistryNotReady("Apps aren't loaded yet.")
    E   django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
    ________________ ERROR collecting tests/panels/test_template.py ________________
    tests/panels/test_template.py:1: in <module>
        from django.contrib.auth.models import User
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/models.py:2: in <module>
        from django.contrib.auth.base_user import AbstractBaseUser, BaseUserManager
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/contrib/auth/base_user.py:47: in <module>
        class AbstractBaseUser(models.Model):
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/db/models/base.py:103: in __new__
        app_config = apps.get_containing_app_config(module)
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:252: in get_containing_app_config
        self.check_apps_ready()
    /nix/store/b6fj23q84lf3w53wpv66kkawmgdrl3g0-python3.9-Django-2.2.24/lib/python3.9/site-packages/django/apps/registry.py:135: in check_apps_ready
        raise AppRegistryNotReady("Apps aren't loaded yet.")
    E   django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
    =========================== short test summary info ============================
    ERROR tests/test_integration.py - django.core.exceptions.AppRegistryNotReady:...
    ERROR tests/commands/test_debugsqlshell.py - django.core.exceptions.AppRegist...
    ERROR tests/panels/test_logging.py - django.core.exceptions.AppRegistryNotRea...
    ERROR tests/panels/test_profiling.py - django.core.exceptions.AppRegistryNotR...
    ERROR tests/panels/test_sql.py - django.core.exceptions.AppRegistryNotReady: ...
    ERROR tests/panels/test_template.py - django.core.exceptions.AppRegistryNotRe...
    !!!!!!!!!!!!!!!!!!! Interrupted: 6 errors during collection !!!!!!!!!!!!!!!!!!!!
    ============================== 6 errors in 1.49s ===============================
    builder for '/nix/store/4r2v0p1lpm4n0zzvmfc25b259k77s13r-python3.9-django-debug-toolbar-3.2.2.drv' failed with exit code 2
    error: build of '/nix/store/4r2v0p1lpm4n0zzvmfc25b259k77s13r-python3.9-django-debug-toolbar-3.2.2.drv' failed

Nobody on NixOS matrix seem to know a fix for this as well, so I just `doCheck = false;`

### CODE REVIEW

**Comment by risicle** [comment]
Have you tried a command along the lines of

```
python manage.py test --settings=parsifal.settings.tests --verbosity=2
```

similar to what the project calls @ https://github.com/vitorfs/parsifal/blob/5c7e902b14fb068e5c1a14f7d55a08f2d40779c7/tox.ini#L14 ?

From the looks of the projects github actions they start up a postgres instance, so it's possible they'll not get very far without one. Worth a try though.

**Comment by risicle** [comment]
Point still stands - upstream seem to invoke their tests like:

```
DJANGO_SETTINGS_MODULE=tests.settings \
		python -m django test $${TEST_ARGS:-tests}
```

as in https://github.com/jazzband/django-debug-toolbar/blob/d779328b19ef9c7835be1c80acedca752a67f3e8/Makefile#L32

You'll probably also want to set environment variables

```
    DB_BACKEND = sqlite3
    DB_NAME = ":memory:"
```

as in https://github.com/jazzband/django-debug-toolbar/blob/d779328b19ef9c7835be1c80acedca752a67f3e8/tox.ini#L61

File: pkgs/development/python-modules/django-debug-toolbar/default.nix [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    tox
+    black
+    pytest
+  ];
+  checkPhase = ''
+    runHook preCheck
+    DJANGO_SETTINGS_MODULE=tests.settings \
+      pytest tests
+    runHook postCheck
+  '';
```
Comment (samuela) [inline pkgs/development/python-modules/django-debug-toolbar/default.nix:68]: All of this can be removed if `doCheck = false`

**Review by samuela — CHANGES_REQUESTED** [review]

**Comment by yuuyins** [comment]
> Point still stands - upstream seem to invoke their tests like:
> 
> ```
> DJANGO_SETTINGS_MODULE=tests.settings \
> 		python -m django test $${TEST_ARGS:-tests}
> ```
> 

thank you! i've changed that to

```
  DB_BACKEND = "sqlite3";
  DB_NAME = ":memory:";
  TEST_ARGS = "tests";

  checkInputs = [
    coverage
    flake8
    html5lib
    isort
    selenium
    tox
    black
    python
    sqlite
  ];
  checkPhase = ''
    runHook preCheck
    DJANGO_SETTINGS_MODULE=tests.settings \
      python -m django test ${TEST_ARGS}
    runHook postCheck
  '';
```

and the tests now went ok. (i tried `\$\$\{TEST_ARGS:-tests}` and variants but was getting some errors, so i just set `${TEST_ARGS}` as `tests`)

```
Finished executing pipInstallPhase
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/dqlpn10slqkqyq5jxd6zfryy3w4fikvk-python3.9-django-debug-toolbar-3.2.2
strip is /nix/store/v8imx1nvyz0hgvx9cbcmh6gp4ngw3ffj-binutils-2.35.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/dqlpn10slqkqyq5jxd6zfryy3w4fikvk-python3.9-django-debug-toolbar-3.2.2/lib
patching script interpreter paths in /nix/store/dqlpn10slqkqyq5jxd6zfryy3w4fikvk-python3.9-django-debug-toolbar-3.2.2
checking for references to /build/ in /nix/store/dqlpn10slqkqyq5jxd6zfryy3w4fikvk-python3.9-django-debug-toolbar-3.2.2...
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
running install tests
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
...............................................ss....s.....s.s...............................s......................sssssss...
----------------------------------------------------------------------
Ran 126 tests in 2.562s

OK (skipped=13)
Destroying test database for alias 'default'...
pythonCatchConflictsPhase
pythonRemoveBinBytecodePhase
pythonImportsCheckPhase
Executing pythonImportsCheckPhase
```

File: pkgs/development/python-modules/django-debug-toolbar/default.nix (hunk 1 above)
Comment (yuuyins) [reply pkgs/development/python-modules/django-debug-toolbar/default.nix]: @samuela the checkPhase now is working.

**Review by samuela — APPROVED** [review]

**Comment by risicle** [comment]
A general tip - if you're having trouble getting a project to build or its tests to work, take a peek at their CI scripts.

**Review by mweinelt — CHANGES_REQUESTED** [review]
That looks pretty solid. A few finishing touches.

File: pkgs/development/python-modules/django-debug-toolbar/default.nix [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    # test
+    ${python.interpreter} -m django test ${TEST_ARGS}
+
+    # coverage
+    ${python.interpreter} --version
+    ${python.interpreter} -b -W always -m coverage run -m django test -v2 ${TEST_ARGS}
+    coverage report
+    coverage html
+    coverage xml
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/django-debug-toolbar/default.nix:56]:
```suggestion
```

File: pkgs/development/python-modules/django-debug-toolbar/default.nix [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  checkPhase = ''
+    runHook preCheck
+
+    # test
+    ${python.interpreter} -m django test ${TEST_ARGS}
+
+    # coverage
+    ${python.interpreter} --version
+    ${python.interpreter} -b -W always -m coverage run -m django test -v2 ${TEST_ARGS}
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/django-debug-toolbar/default.nix:53]:
```suggestion
    ${python.interpreter} -b -W always -m django test -v2 ${TEST_ARGS}
```

**Review by SuperSandro2000 — COMMENTED** [review]
Also please squash the commits.

File: pkgs/development/python-modules/django-debug-toolbar/default.nix (hunk 3 above)
Comment (yuuyins) [reply pkgs/development/python-modules/django-debug-toolbar/default.nix]: i don't understand why to remove the coverage test, but i removed it altogether anyway.

Comment (SuperSandro2000) [reply pkgs/development/python-modules/django-debug-toolbar/default.nix]: We can't upload the coverage results and generating them takes resources so the general consensus is to not generate them in the first place.

Comment (mweinelt) [reply pkgs/development/python-modules/django-debug-toolbar/default.nix]: Not sure what you mean with upload here.

The reason we don't do coverage is that as a metric we don't care for it. Most of the time we are in no position to change coverage of upstream projects.

**Pull request merged**